crates/sui-indexer/migrations/pg/2024-09-12-213234_watermarks/up.sql

```diff
@@ -0,0 +1,22 @@
+CREATE TABLE watermarks
+(
+    -- The table name of group of tables governed by this watermark, i.e `epochs`, `checkpoints`,
+    -- `transactions`.
+	entity                      TEXT          NOT NULL,
+    -- Upper bound epoch range to enable per-entity epoch-level retention policy.
+    epoch_hi                    BIGINT        NOT NULL,
+    -- Lower bound epoch range to enable per-entity epoch-level retention policy.
+    epoch_lo                    BIGINT        NOT NULL,
+    -- The inclusive high watermark that the committer advances.
+	hi                          BIGINT        NOT NULL,
+    -- The inclusive low watermark that the pruner advances. Data before this watermark is
+    -- considered pruned.
+	lo                          BIGINT        NOT NULL,
+    -- Pruner sets this, and uses this column to determine whether to prune or wait long enough that
+    -- all in-flight reads complete or timeout before it acts on an updated watermark.
+	timestamp_ms                BIGINT        NOT NULL,
```

Comment:
What is the semantics of this column? Is it the timestamp whenever a pruning just finished a checkpoint?

Reply:
It should be set by the pruner the last time `lo` was increased, so that the pruner knows to wait until this timestamp + some delta before it prunes up to this new boundary. The delta should be long enough that in flight read requests can land before the pruning starts.

Note that this also means on disaster recovery, the pruner should wait for this timestamp (and delta) to start at all.